### PR TITLE
asserts/snapasserts: introduce ValidationSets

### PR DESCRIPTION
**Pull request by pedronis**
ValidationSets can hold a combination of validation-set assertions.

It will grow code to help applying such combinations.

For now it is able to detect conflicts among them and produce errors representing the conflict.

### CODE REVIEW

File: asserts/validation_set.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ import (
 // Presence represents a presence constraint.
 type Presence string
 
-var (
+const (
```
Comment (pedronis) [inline asserts/validation_set.go:35]: this is a small drive-by that I noticed.

File: asserts/snapasserts/validation_sets.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+	//   presence as invalid
+	revisions map[snap.Revision][]*revConstraint
+}
+
+type revConstraint struct {
+	validationSetKey string
+	asserts.ValidationSetSnap
+}
+
+func (c *snapContraints) conflict() (conflicting []string, err error) {
```
Comment (zyga) [inline asserts/snapasserts/validation_sets.go:78]: This function has unusual signature. I suspect it would be less surprising to return this instead:
```suggestion
func (c *snapContraints) conflict() (conflicting []string, errMsg string) {
```

This way the caller may well expect to return both `conflicting` and `errMsg`.

File: asserts/snapasserts/validation_sets.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+				msg += " or"
+			}
+			if invalid {
+				msg += fmt.Sprintf(" at any revision %s", which)
+			} else {
+				msg += fmt.Sprintf(" required at any revision %s", which)
+			}
+		}
+	}
+	return conflicting, fmt.Errorf(msg)
```
Comment (zyga) [inline asserts/snapasserts/validation_sets.go:137]: Nitpick, while safe for other reasons, `fmt.Errorf` should not be used this way, as `msg` may by accident contain formatting escape codes.

File: asserts/snapasserts/validation_sets.go [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+}
+
+// Conflict returns a non-nil error if the combination is in conflict,
+// nil otherwise.
+func (v *ValidationSets) Conflict() error {
+	sets := make(map[string]*asserts.ValidationSet)
+	snaps := make(map[string]error)
+
+	for snapID, snConstrs := range v.snaps {
+		conflicting, err := snConstrs.conflict()
```
Comment (zyga) [inline asserts/snapasserts/validation_sets.go:228]: If you take my suggestion then this side needs to change as well, and will most likely read more expectedly.

**Review by zyga — APPROVED** [review]
I didn't review the test, as time ran out. Everything looks okay, with the exception of the internal `conflicting` method that I suggested a small tweak for, which does not impact correctness, just being more conventional.

File: asserts/snapasserts/validation_sets.go [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+	}
+	if len(revnos) == 1 {
+		msg += fmt.Sprintf(" at revision %d %s", revnos[0], whichSets(c.revisions[snap.R(revnos[0])], dontCare))
+	} else if len(revnos) > 1 {
+		sort.Ints(revnos)
+		l := make([]string, 0, len(revnos))
+		for _, rev := range revnos {
+			l = append(l, fmt.Sprintf("%d %s", rev, whichSets(c.revisions[snap.R(rev)], dontCare)))
+		}
+		msg += fmt.Sprintf(" at different revisions %s", strings.Join(l, ", "))
```
Comment (stolowski) [inline asserts/snapasserts/validation_sets.go:121]: Random remark (I know we do this sort of error string building already) - this makes i18n impossible (let's hope we don't have to).

**Review by stolowski — APPROVED** [review]
Looks good, thank you!

File: asserts/snapasserts/validation_sets.go (hunk 2 above)
Comment (pedronis) [reply asserts/snapasserts/validation_sets.go]: I did something slightly different in the end, I made conflict() return a snapConflictsError carrying the sets found by whichSets here and deferred some of the message formatting to snapConflictsError.Error

File: asserts/snapasserts/validation_sets.go [hunk 6]
```diff
@@ -75,54 +75,91 @@ type revConstraint struct {
...
+	revisions map[snap.Revision][]string
+}
+
+func (e *snapConflictsError) Error() string {
+	whichSets := func(which []string) string {
 		return fmt.Sprintf("(%s)", strings.Join(which, ","))
 	}
 
-	msg := fmt.Sprintf("cannot constraint snap %q", c.name)
+	msg := fmt.Sprintf("cannot constraint snap %q", e.name)
```
Comment (stolowski) [inline asserts/snapasserts/validation_sets.go:137]: Just noticed this now, should this be "cannot constrain ..." (no "t")?

**Review by stolowski — APPROVED** [review]
The last commit looks fine to me. One remark about error message that I missed previously.

File: asserts/snapasserts/validation_sets.go (hunk 6 above)
Comment (pedronis) [reply asserts/snapasserts/validation_sets.go:137]: oops, thanks
